### PR TITLE
Add weighted random sampler

### PR DESCRIPTION
**Pull request by heavengate**
### PR types
New features

### PR changes
APIs

### Describe
add `paddle.io.WeightedRandomSampler`
WeightedRandomSampler_en

### CODE REVIEW

**Review by LielinJiang — DISMISSED** [review]

File: python/paddle/fluid/dataloader/sampler.py [hunk 1]
```diff
@@ -234,3 +238,85 @@ def __iter__(self):
...
         return self.num_samples
+
+
+def _weighted_sample(probs, num_samples, replacement=True):
+    if isinstance(probs, core.LoDTensor):
+        probs = probs.numpy()
+    if isinstance(probs, (list, tuple)):
+        probs = np.array(probs)
+    assert isinstance(probs, np.ndarray), \
+            "probs should be paddle.Tensor, numpy.ndarray, list or tuple"
```
Comment (qingqing01) [inline python/paddle/fluid/dataloader/sampler.py]: 对用户的接口是用weights，报错信息和接口arg保持一致

File: python/paddle/fluid/dataloader/sampler.py [hunk 2]
```diff
@@ -234,3 +238,85 @@ def __iter__(self):
...
+    for i in range(probs.shape[0]):
+        ret = np.random.choice(probs.shape[1], num_samples, replacement,
+                               probs[i])
+        rets.append(ret)
+    return np.array(rets)
+
+
+class WeightedRandomSampler(Sampler):
+    """
+    Random sample with give weights (probabilities), sampe index will be in range
```
Comment (qingqing01) [inline python/paddle/fluid/dataloader/sampler.py]: with give  -> with given

File: python/paddle/fluid/dataloader/sampler.py (hunk 1 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/sampler.py]: Done, thanks!

File: python/paddle/fluid/dataloader/sampler.py (hunk 2 above)
Comment (heavengate) [reply python/paddle/fluid/dataloader/sampler.py]: Done, thanks!

**Review by saxon-zh — DISMISSED** [review]
LGTM

**Review by qingqing01 — APPROVED** [review]

**Review by saxon-zh — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**